async-iterator.md

```diff
@@ -0,0 +1,904 @@
+
+I'll share some notes about the design for async-iterator methods (C# 8.0 feature).
+
+Editorial notes:
+- Some sections probably should be skipped on first reading. I marked them as "optional".
+- Although I will gloss over many important parts, I will try to mention them to give a sense of how things fit together.
+- `/* ... */` comments indicate omitted code, while `// ... ` comments apply to code that is there.
+- Every illustration of generated code omits some details.
+
+
+## Overview
+
+Async-iterator methods are methods with both yields and awaits and produce async-enumerables (`IAsyncEnumerable<T>`) or async-enumerators (`IAsyncEnumerator<T>`).
+Those are methods that yield a sequence of items (like iterator methods that return `IEnumerable<T>`), but allow for asynchronous computation (like async methods).
+
+
+## Consumption with `await foreach`
+
+The `await foreach` statement offers the most straightforward way of enumerating an async-enumerable's items.
+It is very similar to `foreach` for enumerables. The main difference is that `await foreach` awaits each item (and also disposal).
+
+The parallel jumps out in the interfaces involved:
+
+| async enumerables | enumerables |
+| --- | --- |
+| `IAsyncEnumerable<T>` used with `GetAsyncEnumerator()` | `IEnumerable<T>` used with `GetEnumerator()` |
+| `IAsyncEnumerator<T>` used with `await MoveNextAsync()` and `Current` | `IEnumerator<T>` used with `MoveNext()` and `Current` |
+| `IAsyncDisposable` used with `await DisposeAsync()` | `IDisposable` used with `Dispose()` |
+
+So it is unsurprising that the lowering for `await foreach` is very similar to that of a `foreach`.
+
+The following:
+```C#
+await foreach (var value in asyncEnumerable)
+{
+   ... body using `value` ...
+}
+```
+translates to:
+```C#
+var enumerator = asyncEnumerable.GetAsyncEnumerator();
+try
+{
+    while (await enumerator.MoveNextAsync())
+    {
+        var value = enumerator.Current;
+        ... body using `value` ...
+    }
+}
+finally
+{
+    await enumerator.DisposeAsync();
+}
+```
+
+
+### Control flow (caller and background execution)
+
+In a `await foreach`, the caller repeatedly calls `MoveNextAsync()`.
+This returns a `ValueTask<bool>` where the `bool` represents the presence or absence of an item.
+When you call `MoveNextAsync()`:
+- if the code reaches a `yield return` statement, you immediately get a `true`,
+- if the code reaches the end of the method, you immediately get a `false`,
+- if you reach an `await`, you get a pending task and the code continues to execute in the background.
+
+As the code continues to execute in the background:
+- if it reaches a `yield return`, the background execution completes and the task we've previously handed to the caller is fulfilled with `true`,
+- if it reaches the end of the method, that task is fulfilled with `false`,
+- if it reaches an `await`, execution continues in the background.
```

Comment:
This last bullet isn't really necessary.  From the caller's perspective, this isn't visible, separate from the lifecycle of the `ValueTask<bool>` that's returned to them.